### DISCUSSION TITLE
How to use Renderer.Blit?

### DISCUSSION THREAD

**Discussion by jackdaus** [StereoKit/StereoKit#533] [Q&A] [answered]
I'm still having a _blit_ of trouble figuring out how to use Renderer.Blit. I'm moving this convo to GitHub so other folks can find it for reference more easily. 

Below is some code to minimally reproduce what I'm trying to do:

``` csharp
using StereoKit;
using System;

// Initialize StereoKit
SKSettings settings = new SKSettings
{
	appName = "BlitDemo",
	assetsFolder = "Assets",
};
if (!SK.Initialize(settings))
	Environment.Exit(1);

// Create assets
Material _sourceMaterial = Default.MaterialUnlit.Copy();
_sourceMaterial[MatParamName.DiffuseTex] = Tex.DevTex;

Tex _renderTex = new Tex(TexType.Rendertarget);
_renderTex.SetSize(500, 500);
_renderTex.AddZBuffer(TexFormat.Depth32);

Material _renderTexMaterial = Default.MaterialUnlit.Copy();
_renderTexMaterial[MatParamName.DiffuseTex] = _renderTex;
_renderTexMaterial.FaceCull = Cull.None;

Model _blitModel = Model.FromMesh(
	Mesh.Quad,
	_renderTexMaterial);

// Core application loop
while (SK.Step(() =>
{
	Renderer.Blit(_renderTex, _sourceMaterial);

	_blitModel.Draw(Matrix.T(0, 0, -1));
})) ;
SK.Shutdown();
```

I'm expecting the `Tex.DevTex` tex to show up on the quad mesh, but instead I just get black. ⬛

![image](https://user-images.githubusercontent.com/53974525/203341777-309335a7-2cdc-46d6-bb7f-66786f355746.png)

This is my first time using a Blit function, so I'm probably just missing something!

**Comment by maluoi** [comment, marked as answer]
This seems basically right! The only remaining thing is that you will need a special shader for this to work. `Material.Unlit` still includes transformation matrix logic that requires a camera present, where Blit is a much simpler sort of rendering environment that has no camera!

The preview builds do include a builtin blit shader, which hasn't made it conveniently into the C# layer just yet! You can grab it like this right now in a recent v0.3.7-preview build, I'd recommend preview.5, preview.6 has some shader build issues.

```csharp
Material blitMat = new Material(Shader.Find("default/shader_blit"));
```

You can grab this shader [source over here](https://github.com/StereoKit/StereoKit/blob/develop/StereoKitC/shaders_builtin/shader_builtin_blit.hlsl) and use it as a template for your own blit shaders! And when v0.3.7 comes out, this shader should be available as `Material.Blit` or something similar.

**Reply by jackdaus** [reply]
Awesome, thank you! I got it working with that shader. (And as an aside, it's working for me with v0.3.7-preview.6. But I'm not using any other shaders beyond this built-in one.)

And just in case someone else needs it, this is the code to do a simple blit (as of v0.3.7-preview.6)

``` csharp
using StereoKit;
using System;

// Initialize StereoKit
SKSettings settings = new SKSettings
{
	appName = "BlitDemo",
	assetsFolder = "Assets",
};
if (!SK.Initialize(settings))
	Environment.Exit(1);

// Create assets
Material _blitMat = new Material(Shader.Find("default/shader_blit"));
_blitMat["source"] = Tex.DevTex;

Tex _renderTex = new Tex(TexType.Rendertarget);
_renderTex.SetSize(500, 500);
_renderTex.AddZBuffer(TexFormat.Depth32);

Material _renderTexMaterial = Default.MaterialUnlit.Copy();
_renderTexMaterial[MatParamName.DiffuseTex] = _renderTex;
_renderTexMaterial.FaceCull = Cull.None;

Model _blitModel = Model.FromMesh(
	Mesh.Quad,
	_renderTexMaterial);

Renderer.Blit(_renderTex, _blitMat);

// Core application loop
while (SK.Step(() =>
{
	_blitModel.Draw(Matrix.T(0, 0, -1));
})) ;
SK.Shutdown();
```
